### PR TITLE
NEWRELIC-6167 Update `getBrowserTimingHeader` to inject Browser Agent regardless of transaction availability

### PR DESCRIPTION
**Pull request by jmartin4563**
## Proposed Release Notes
* Update `getBrowserTimingHeader` to allow Browser Agent to be generated even when not in a Transaction by adding `allowTransactionlessInjection` to function options. `allowTransactionlessInjection` is a boolean option, and when set to `true`, will allow injection of the Browser Agent when not in a transaction. Note that if you are using this option, you may need to wait until the Node agent has established a connection before calling `getBrowserTimingHeader`. To wait until the agent is connected, you can add the following check to your code: 
```js
if (!newrelic.agent.collector.isConnected()) {
  await new Promise((resolve) => {
    newrelic.agent.on('connected', resolve)
  })
}
```

## Links
Closes https://github.com/newrelic/newrelic-node-nextjs/issues/99

## Details
This has been a source of pain for users in a SSR/SSG applications, where they may not necessarily have an active transaction when injecting the Browser Agent. Before this change, in these scenarios the Browser Agent would just not be injected, and users would have to fall back to the Copy/Paste method, which is hard to maintain over time. After this change, the Browser Agent will be injected, but will be missing the transaction related information. It's also worth noting that in this situation, users won't be able to do custom attribute forwarding, as it's reliant on transaction state.

### CODE REVIEW

File: api.js [hunk 1]
```diff
@@ -640,152 +640,177 @@ function _generateRUMHeader(options = {}, metadata, loader) {
...
+    if (!properties.isEmpty(agentAttrs)) {
+      attrs.a = agentAttrs
+    }
+
+    if (!properties.isEmpty(attrs)) {
+      rumHash.atts = hashes.obfuscateNameUsingKey(JSON.stringify(attrs), key)
+    }
+  } else {
+    logger.debugOnce(
+      'api:getBrowserTimingHeader',
```
Comment (jordigh) [inline api.js:803]: nit: would it really be that noisy to use plain `logger.debug` here? I understand server-side rendering is reasonably static and you wouldn't be doing lots of attempts to render without a transaction.

On the other hand, if you accidentally try to inject the header as normal but you're accidentally not in a transaction, getting debug output every time could be helpful.

File: test/unit/rum.test.js [hunk 2]
```diff
@@ -49,10 +49,6 @@ describe('the RUM API', function () {
     api.getBrowserTimingHeader().should.equal('<!-- NREUM: (0) -->')
   })
 
-  it('should issue a warning outside a transaction', function () {
-    api.getBrowserTimingHeader().should.equal('<!-- NREUM: (1) -->')
-  })
-
```
Comment (jordigh) [inline test/unit/rum.test.js:55]: Do we still have a test for `_graceFail(1)`?

**Review by jordigh — COMMENTED** [review]
I like that all of the error-checking has been now factored out into its own function.

The change makes sense: if there's a transaction, just add all of the transaction info into the hash, otherwise, just hash with what you have.

File: test/unit/rum.test.js (hunk 2 above)
Comment (jmartin4563) [reply test/unit/rum.test.js:55]: Yessir, on line 52:
```
  it('should issue a warning if the transaction was ignored', function () {
    helper.runInTransaction(agent, function (tx) {
      tx.ignore = true
      api.getBrowserTimingHeader().should.equal('<!-- NREUM: (1) -->')
    })
  })
```

**Review by bizob2828 — COMMENTED** [review]
I tested this with our sample next.js app(PR pending to add support for SSG), but this solution doesn't work. it returns with `NREUM(4)`. I suspect this is because the agent doesn't connect fast enough at build time.  This might just be a tweak to how to inject browser agent on static pages so I'll keep digging.  I'm also wondering if we want to wrap no longer requiring a transaction in an option passed to getBrowserTimingHeader.  I don't know the affects this may have on our existing customer base.  Were you able to take a look to see how other languages agents handle this?

**Comment by bizob2828** [comment]
Ok I confirmed this solution does work on static pages. I just had to add a promise to wait for agent to be connected first. https://github.com/newrelic-experimental/newrelic-nextjs-integration/pull/9

File: api.js (hunk 1 above)
Comment (bizob2828) [reply api.js:803]: yea this won't matter.  I've noticed in the SSG case it spins up separate threads so the debug once is useless unfortunately

**Comment by jmartin4563** [comment]
> I tested this with our sample next.js app(PR pending to add support for SSG), but this solution doesn't work. it returns with `NREUM(4)`. I suspect this is because the agent doesn't connect fast enough at build time. This might just be a tweak to how to inject browser agent on static pages so I'll keep digging. I'm also wondering if we want to wrap no longer requiring a transaction in an option passed to getBrowserTimingHeader. I don't know the affects this may have on our existing customer base. Were you able to take a look to see how other languages agents handle this?

I looked at Ruby/Python/.NET/Go, and all of them require that a transaction context be available for their equivalent functions to work, so we'll be the different one in this scenario. To your point about adding an option to the function, I agree. Additionally, that might be something that can be agent-spec'd if we find that it's a useful thing in other agents.

**Comment by bizob2828** [comment]
> > I tested this with our sample next.js app(PR pending to add support for SSG), but this solution doesn't work. it returns with `NREUM(4)`. I suspect this is because the agent doesn't connect fast enough at build time. This might just be a tweak to how to inject browser agent on static pages so I'll keep digging. I'm also wondering if we want to wrap no longer requiring a transaction in an option passed to getBrowserTimingHeader. I don't know the affects this may have on our existing customer base. Were you able to take a look to see how other languages agents handle this?
> 
> I looked at Ruby/Python/.NET/Go, and all of them require that a transaction context be available for their equivalent functions to work, so we'll be the different one in this scenario. To your point about adding an option to the function, I agree. Additionally, that might be something that can be agent-spec'd if we find that it's a useful thing in other agents. Did you have any thoughts on the option name?

You may want to figure out if other teams can recall why they built this function to behave like this. We don't have a true spec for this as well just examples on what the payload should look like

**Review by bizob2828 — APPROVED** [review]

**Comment by bizob2828** [comment]
I would add more to the proposed release notes. especially since you may need to add more logic to your code to wait for agent to be connected before calling this.  You can see [here](https://github.com/newrelic-experimental/newrelic-nextjs-integration/pull/9/files#diff-4aaf885cfc46070419a446e27afeee17f0aacb25a188f74ef7dac690424f8a71R24) I added a check to see if collector is connected and if not wait for it to be connected

**Review by bizob2828 — APPROVED** [review]

**Pull request merged**